crowbar_framework/app/models/proposal.rb

```diff
@@ -0,0 +1,118 @@
+class Proposal < ActiveRecord::Base
+  include Crowbar::ProposalMethods
+
+  # FIXME: remove this
+  class_attribute :chef_type
+
+  self.chef_type = "data_bag_item"
+
+  # FIXME: add proper i18n to errors
+
+  class TemplateMissing < StandardError; end
+  class TemplateInvalid < StandardError; end
+
+  serialize :properties, JSON
+
+  validates :name, :barclamp, :properties, presence: true
+  validate  :name, :name_not_on_blacklist
+  validates :name, uniqueness: { scope: :barclamp }
+
+  after_initialize :load_properties_template
+  before_save      :update_proposal_id
+
+  # XXX: a 'registered' barclamp could have a has_many :proposals and have a factory
+  # method for creating them. Then the check for barclamp arg would not be needed,
+  # as we'd always know the barclamp exists.
+  def initialize(attributes = nil, options = {})
+    raise ArgumentError.new("Barclamp attribute is required") unless attributes && attributes.key?(:barclamp)
+
+    super
+  end
+
+  def self.find_proposals(barclamp)
+    where(:barclamp => barclamp)
+  end
+
+  def self.find_barclamp(barclamp)
+    self.new(:barclamp => barclamp, :name => "template")
```

Comment:
> > +  def self.find_barclamp(barclamp)
> > +    self.new(:barclamp => barclamp, :name => "template")
> 
> Does that mean that we put the template in the database? Do we really need to do it?

This just creates a new object, but does not save it (until the caller
does that, of course).